Source/ConvolutionFunctions/arm_convolve_1_x_n_s4.c

```diff
@@ -0,0 +1,211 @@
+ * Project:      CMSIS NN Library
+ * Title:        arm_convolve_1_x_n_s4.c
+ * Description:  s4 version of 1xN convolution using symmetric quantization.
+ *
+ * $Date:        10 April 2024
+ * $Revision:    V.1.0.0
+ *
+ * Target :  Arm(R) M-Profile Architecture
+ *
+ * -------------------------------------------------------------------- */
+
+#include "arm_nnfunctions.h"
+#include "arm_nnsupportfunctions.h"
+/**
+ *  @ingroup Public
+ */
+
+/**
+ * @addtogroup NNConv
+ * @{
+ */
+
+/*
+ * 1xN s4 convolution function.
+ *
+ * Refer header file for details.
+ *
+ */
+
+arm_cmsis_nn_status arm_convolve_1_x_n_s4(const cmsis_nn_context *ctx,
+                                          const cmsis_nn_conv_params *conv_params,
+                                          const cmsis_nn_per_channel_quant_params *quant_params,
+                                          const cmsis_nn_dims *input_dims,
+                                          const int8_t *input_data,
+                                          const cmsis_nn_dims *filter_dims,
+                                          const int8_t *filter_data,
+                                          const cmsis_nn_dims *bias_dims,
+                                          const int32_t *bias_data,
+                                          const cmsis_nn_dims *output_dims,
+                                          int8_t *output_data)
+{
+    arm_cmsis_nn_status status = ARM_CMSIS_NN_SUCCESS;
+    int32_t buffer_size = arm_convolve_1_x_n_s4_get_buffer_size(conv_params, input_dims, filter_dims, output_dims);
+    /* The wrapper API is the ultimate reference for argument check */
+    if ((input_dims->h != 1) || conv_params->dilation.w != 1 || (buffer_size != 0 && ctx->buf == NULL) ||
+        conv_params->stride.w == 0 || (conv_params->stride.w * input_dims->c % 4 != 0))
+    {
+        status = ARM_CMSIS_NN_ARG_ERROR;
+        goto out;
+    }
+
+#if defined(ARM_MATH_MVEI)
+    (void)bias_dims;
+    const uint16_t input_x = input_dims->w;
+    const uint16_t kernel_x = filter_dims->w;
+    const uint16_t output_x = output_dims->w;
+    const uint16_t output_ch = output_dims->c;
+    const uint16_t input_ch = input_dims->c;
+    const uint16_t pad_x = conv_params->padding.w;
+    const uint16_t stride_x = conv_params->stride.w;
+
+    // Total pad for dilation of 1
+    const int32_t total_pad = ((output_x - 1) * stride_x + kernel_x - input_x);
+    const int32_t asym_pad = total_pad % 2;
+
+    if (pad_x * 2 + asym_pad != total_pad)
+    {
+        return ARM_CMSIS_NN_FAILURE;
+    }
+
+    const int32_t right_pad_num = pad_x + asym_pad != 0 ? MAX(1, (pad_x + asym_pad + stride_x - 1) / stride_x) : 0;
+    const int32_t left_pad_num = pad_x != 0 ? MAX(1, (pad_x + stride_x - 1) / stride_x) : 0;
+    const int32_t no_pad_num = MAX(output_x - (right_pad_num + left_pad_num), 0);
+
+    if (right_pad_num + no_pad_num + left_pad_num != output_x)
```

Comment:
could you explain the case where the padded and unpadded elements do not add up to output_x?

Reply:
This is a copy of the int8 version that fixes this issue: https://github.com/ARM-software/CMSIS-NN/issues/109

Reply:
We will handle this in a separate patch